internal/datastore/memdb/readwrite.go

```diff
@@ -0,0 +1,210 @@
+package memdb
+
+import (
+	"fmt"
+
+	v1 "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	"github.com/hashicorp/go-memdb"
+	"github.com/jzelinskie/stringz"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/authzed/spicedb/pkg/datastore"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+type memdbReadWriteTx struct {
+	memdbReader
+	newRevision datastore.Revision
+}
+
+func (rwt *memdbReadWriteTx) WriteRelationships(mutations []*v1.RelationshipUpdate) error {
+	rwt.lockOrPanic()
+	defer rwt.Unlock()
+
+	tx, err := rwt.txSource()
+	if err != nil {
+		return err
+	}
+
+	return rwt.write(tx, mutations)
+}
+
+// Caller must already hold the concurrent access lock!
+func (rwt *memdbReadWriteTx) write(tx *memdb.Txn, mutations []*v1.RelationshipUpdate) error {
+	// Apply the mutations
+	for _, mutation := range mutations {
+		rel := &relationship{
+			mutation.Relationship.Resource.ObjectType,
+			mutation.Relationship.Resource.ObjectId,
+			mutation.Relationship.Relation,
+			mutation.Relationship.Subject.Object.ObjectType,
+			mutation.Relationship.Subject.Object.ObjectId,
+			stringz.DefaultEmpty(mutation.Relationship.Subject.OptionalRelation, datastore.Ellipsis),
+		}
+
+		found, err := tx.First(
+			tableRelationship,
+			indexID,
+			rel.namespace,
+			rel.resourceID,
+			rel.relation,
+			rel.subjectNamespace,
+			rel.subjectObjectID,
+			rel.subjectRelation,
+		)
+		if err != nil {
+			return fmt.Errorf("error loading existing relationship: %w", err)
+		}
+
+		var existing *relationship
+		if found != nil {
+			existing = found.(*relationship)
+		}
+
+		switch mutation.Operation {
+		case v1.RelationshipUpdate_OPERATION_CREATE:
+			if existing != nil {
+				return fmt.Errorf("duplicate relationship found for create operation")
+			}
+			fallthrough
+		case v1.RelationshipUpdate_OPERATION_TOUCH:
+			if err := tx.Insert(tableRelationship, rel); err != nil {
+				return fmt.Errorf("error inserting relationship: %w", err)
+			}
+		case v1.RelationshipUpdate_OPERATION_DELETE:
+			if existing != nil {
+				if err := tx.Delete(tableRelationship, existing); err != nil {
+					return fmt.Errorf("error deleting relationship: %w", err)
+				}
+			}
```

Comment:
We don't fail if you try to delete something that doesn't exist? I guess that makes sense based on the SQL expressions for other datastores.